DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEventFlat.h

```diff
@@ -0,0 +1,180 @@
+/// \author pietro.cortese@cern.ch
+
+namespace o2
+{
+namespace zdc
+{
+using FirstEntry = int;
+using NElem = int;
+
+struct RecEventFlat {
+  o2::InteractionRecord ir;
+  uint32_t channels = 0;                      /// pattern of channels acquired
+  uint32_t triggers = 0;                      /// pattern of channels with autotrigger bit
+  std::map<uint8_t, float> ezdc;              /// signal in ZDCs
+  std::vector<int16_t> TDCVal[NTDCChannels];  /// TdcChannels
+  std::vector<int16_t> TDCAmp[NTDCChannels];  /// TdcAmplitudes
+  std::vector<o2::zdc::BCRecData>* mRecBC;    //! Interaction record and references to data
+  std::vector<o2::zdc::ZDCEnergy>* mEnergy;   //! ZDC energy
+  std::vector<o2::zdc::ZDCTDCData>* mTDCData; //! ZDC TDC
+  std::vector<uint16_t>* mInfo;               //! Event quality information
+  uint64_t mEntry = 0;                        //! Current entry
+  uint64_t mNEntries = 0;                     //! Number of entries
+  FirstEntry mFirstE = 0;                     //! First energy
+  FirstEntry mFirstT = 0;                     //! First TDC
+  FirstEntry mFirstI = 0;                     //! First info
+  FirstEntry mStopE = 0;                      //! Last + 1 energy
+  FirstEntry mStopT = 0;                      //! Last + 1 TDC
+  FirstEntry mStopI = 0;                      //! Last + 1 info
+  NElem mNE = 0;                              //! N energy
+  NElem mNT = 0;                              //! N TDC
+  NElem mNI = 0;                              //! N info
+  o2::zdc::BCRecData mCurB;                   //! Current BC
+  std::array<bool, NChannels> tdcPedEv;       /// Event pedestal for TDC
+  std::array<bool, NChannels> tdcPedOr;       /// Orbit pedestal for TDC
+  std::array<bool, NChannels> tdcPedQC;       /// QC pedestal for TDC
+  std::array<bool, NChannels> tdcPedMissing;  /// Missing pedestal for ADC
+  std::array<bool, NChannels> adcPedEv;       /// Event pedestal for ADC
+  std::array<bool, NChannels> adcPedOr;       /// Orbit pedestal for ADC
+  std::array<bool, NChannels> adcPedQC;       /// QC pedestal for ADC
+  std::array<bool, NChannels> adcPedMissing;  /// Missing pedestal for ADC
+  uint8_t mVerbosity = DbgZero;               //! Verbosity level
+  uint32_t mTriggerMask = 0;                  //! Trigger mask for printout
+
+  void init(std::vector<o2::zdc::BCRecData>* RecBC, std::vector<o2::zdc::ZDCEnergy>* Energy, std::vector<o2::zdc::ZDCTDCData>* TDCData, std::vector<uint16_t>* Info);
+
+  int next();
+
+  inline NElem getNEnergy()
+  {
+    return mNE;
+  }
+
+  inline NElem getNTDC()
+  {
+    return mNT;
+  }
+
+  inline NElem getNInfo()
+  {
+    return mNI;
+  }
+
+  float tdcV(uint8_t ich, uint64_t ipos)
+  {
+    if (ich < NTDCChannels) {
+      if (ipos < TDCVal[ich].size()) {
+        return FTDCVal * TDCVal[ich][ipos];
+      }
+    }
+    return -std::numeric_limits<float>::infinity();
```

Comment:
how are you later going to process the `infinity` values?

Reply:
In this case it should be QC. Infinity should never happen, unless there is some memory corruption or a bug in the code.
If you use TreePlayer in root it also helps in producing a correct histogram by putting all missing entries in underflow